In [1]:
import pandas as pd
import tarfile
import csv

tar = tarfile.open("/Users/gavincleach/Desktop/famous-chord-progressions-raw-data/billboard-2.0-salami_chords.tar.gz", "r:gz")
famous_chord_progressions = []
for member in tar.getmembers():
    f = tar.extractfile(member)
    famous_chord_progressions.append(f.read())

billboard_charts = pd.read_csv('/Users/gavincleach/Desktop/famous-chord-progressions-raw-data/billboard-2.0-index.csv')
        
len(famous_chord_progressions)

890

In [2]:
billboard_charts.set_index("id", inplace = True)

billboard_charts.head()

chart_date  target_rank  actual_rank                title  \
id                                                              
1   1987-07-11           82          NaN                  NaN   
2   1973-08-04           14          NaN                  NaN   
3   1961-07-03           56         57.0         I Don't Mind   
4   1971-08-07           32         31.0  You've Got A Friend   
5   1976-07-10           51          NaN                  NaN   

                          artist  peak_rank  weeks_on_chart  
id                                                           
1                            NaN        NaN             NaN  
2                            NaN        NaN             NaN  
3                    James Brown       47.0             8.0  
4   Roberta Flack,Donny Hathaway       29.0            12.0  
5                            NaN        NaN             NaN

let's first get rid of the rows in `billboard_charts` that have `NaN` value in the `title` column. There are 890 songs with full chord progressions, so we will see how many entries we have in `billboard_charts` after we drop those rows.

In [3]:
unwanted_portion = billboard_charts.loc[billboard_charts['title'].isnull()]
ids_to_drop = list(unwanted_portion.index) 

billboard_charts = billboard_charts.drop(labels = ids_to_drop, axis = 0)

billboard_charts.shape

(890, 7)

Great! So now we only have the remaining song entries that don't have `NaN` entries.

Before we clean up `billboard_charts`, we should clean up the `famous_chord_progressions` list, because there is much more work to do on that.

It would be beneficial to understand what each item in the list `famous_chord_progressions` means. So...

(ignore \'s)

\# title: \
after this comes the title of the song. 

\# artist: \
after this comes the artist name.

\# metre: \
after this comes the time signature.

\# tonic: \
after this comes the key the song starts in. 

for tonic and metre there my not be a definite answer for each of these, in that case, there will be a '?' in place. Tonic and metre might change throughout the song as well.

In [17]:
import re

james_brown_i_dont_mind = famous_chord_progressions[0]
part_of_james_brown_i_dont_mind = "# title: I Don't mind\n# artist: Jam"

test_str_0 = 'my favorite numbers are 16 and 2\n'
#test_str_1 = 


title_pattern = "# title: (.+)\n"

#artist_pattern = 

#metre_pattern = 

#tonic_pattern = 

result = re.findall(title_pattern, part_of_james_brown_i_dont_mind)

print(result)

["I Don't mind"]
